README.md
KeePassX with password database sharing
=======================================

These changes allow multiple KeePassX instances to open same file (e.g. on a
network share) for reading.

All the changes are done against stock keepassx 0.4.3.

## What is done

 * Don't keep database file open, close it just after loading.
 * Option to open database file in read-only mode.
 * Disabled "Save" button in read-only mode.

## TODO

 * Disable all database changing actions in the GUI in read-only mode.


debian/changelog
keepassx (0.4.3+dfsg-0.1ubuntu1sharing4) raring; urgency=low

  * Added database sharing patch revision 4

 -- Dmitry Frolov (Launchpad.net) <dmitry.frolov@gmail.com>  Mon, 18 Feb 2013 17:54:29 +0700

keepassx (0.4.3+dfsg-0.1ubuntu1) raring; urgency=low

  * Merge from Debian unstable, remaining changes:
    - Build-depend on libmagickcore-extra instead of librsvg2-bin so
      imagemagick correctly renders svg images. (LP: #979340)
    - Add check_systray_available.diff:
      - Check if the systray is available before using it.
      - Disable the tray icon when running on Unity. (LP: #842224)

 -- Felix Geyer <debfx@ubuntu.com>  Tue, 02 Apr 2013 17:44:07 +0200

keepassx (0.4.3+dfsg-0.1) unstable; urgency=high

  * Non-maintainer upload.
  * Use repackaged upstream tarball by Felix Geyer, which replaces icons of
    unknown origin (clientic.png text_block.png text_bold.png text_center.png
    text_italic.png text_left.png text_right.png text_under.png) and adds full
    text of associated licenses (LICENSE.LGPL-2.1 LICENSE.LGPL-3).
  * Use patch by Felix Geyer to update debian/copyright with the origin and
    license of png files, and update README.source accordingly.
  * Additionally, update debian/copyright to document the origin and copyright
    of go-*.png and help_about.png icons (project Oxygen, and license CC-BY-SA
    3.0 or LGPL respectively).
  * The above three items close #698832
  * Update debian/watch for version mangling.
  * Setting high urgency due to RC bug.

 -- Serafeim Zanikolas <sez@debian.org>  Wed, 20 Mar 2013 14:04:39 +0100

keepassx (0.4.3-2ubuntu1) raring; urgency=low

  * Merge from Debian unstable. (LP: #1094568) Remaining changes:
    - Add check_systray_available.diff
    - Disable the tray icon when running on Unity. (LP: #842224)

 -- Alessandro Losavio <alessandro.losavio@yahoo.com>  Sun, 30 Dec 2012 09:39:20 +0000

keepassx (0.4.3-2) unstable; urgency=low

  * Fix FTBFS with gcc 4.7. (Closes: #667224)
    - Add fix_ftbfs_gcc47.diff
  * Bump Standards-Version to 3.9.3, no changes needed.
  * Enable parallel building.
  * Switch to debhelper compat level v9.
    - Exports build flags.
  * Use the qmake_qt4 dh buildsystem.
  * Install application icon in various resolutions.
    - Add debian/keepassx.svg
    - Build-depend on imagemagick and librsvg2-bin.

 -- Felix Geyer <debfx-pkg@fobos.de>  Thu, 05 Apr 2012 13:10:22 +0200

keepassx (0.4.3-1ubuntu3) quantal; urgency=low

  * import fix_ftbfs_gcc47.diff from debian

 -- Julian Taylor <jtaylor@ubuntu.com>  Fri, 14 Sep 2012 20:00:08 +0200

keepassx (0.4.3-1ubuntu2) precise; urgency=low

  * Build-depend on libmagickcore-extra instead of librsvg2-bin so imagemagick
    correctly renders svg images. (LP: #979340)
  * Disable the tray icon when running on Unity. (LP: #842224)

 -- Felix Geyer <debfx@ubuntu.com>  Mon, 16 Apr 2012 12:14:22 +0200

keepassx (0.4.3-1ubuntu1) natty; urgency=low

  * Install application icon in various resolutions. (LP: #501966)
    - Add debian/keepassx.svg
    - Build-depend on imagemagick and librsvg2-bin.
  * Check if the systray is available before using it.
    - Add check_systray_available.diff

 -- Felix Geyer <debfx-pkg@fobos.de>  Sat, 23 Apr 2011 00:31:42 +0200

keepassx (0.4.3-1) unstable; urgency=low

  * New upstream release. (Closes: #572614)
    - Fixes FTBFS with binutils-gold. (Closes: #554982)
  * Switch to new source format 3.0 (quilt).
  * Bump Standards-Version to 3.8.4, no changes needed.

 -- Felix Geyer <debfx-pkg@fobos.de>  Sun, 07 Mar 2010 00:42:51 +0100

keepassx (0.4.1-1) unstable; urgency=low

  * New upstream release. (LP: #431798)
  * Bump debhelper to dh 7.
    - increase debian/compat to 7.
    - simplify debian/rules.
  * Bump Standards-Version to 3.8.3, no changes needed.

 -- Felix Geyer <debfx-pkg@fobos.de>  Thu, 17 Sep 2009 15:21:00 +0200

keepassx (0.4.0-1) unstable; urgency=low

  * New upstream release. (Closes: #522488)
  * adapt debian/watch file to new upstream naming style.

 -- Reinhard Tartler <siretart@tauware.de>  Sat, 04 Apr 2009 11:39:24 +0200

keepassx (0.3.4-2) unstable; urgency=low

  [ Felix Geyer ]
  * introduce quilt
  * Bug fix: "Keepassx crashes on save when password entry changes".
    LP: #311359
    - add debian/patches/svn281_fix_buffer_too_small.patch
    - build-depend on quilt

  [ Reinhard Tartler ]
  * target unstable
  * convert all format patches to quilt
  * add a README.source file documenting quilt usage
  * bump Standards-Version to 3.8.0

 -- Reinhard Tartler <siretart@tauware.de>  Fri, 03 Apr 2009 16:21:55 +0200

keepassx (0.3.4-1) experimental; urgency=low

  * new upstream release. LP: #296014

 -- Reinhard Tartler <siretart@tauware.de>  Sun, 09 Nov 2008 21:59:02 +0100

keepassx (0.3.3-1) experimental; urgency=low

  * new upstream release. Closes: #492423 
  * add Vcs-Bzr header.
  * build and install translations. Closes: #475073
  * add galician translation. Thanks to damufo <damufo@gmail.com> 
    Closes: #499097
  * install files using dh_install instead of 'cp'
  * no longer install our own .desktop file, the one provided by upstream
    is just fine, and rename manpage keepass.1 -> keepassx.1. Closes: #474455
  * remove useless piece of documentation. Closes: #475126
  * tighten dependency on libqt4, drop libpng3-dev dependency. Closes: #474387

 -- Reinhard Tartler <siretart@tauware.de>  Thu, 02 Oct 2008 10:34:53 +0200

keepassx (0.3.3-0ubuntu1) intrepid; urgency=low

  * New upstream release (LP: #250299). 

 -- Nick Ellery <nick.ellery@ubuntu.com>  Fri, 26 Sep 2008 16:48:37 -0700

keepassx (0.3.1-1) unstable; urgency=low

  * Bug fix: "keepassx: Please use better watch file expression", thanks
    to Roberto C. Sanchez (Closes: #474335).
  * Bug fix: "New upstream version 0.3.0a", thanks to Felix Geyer (Closes:
    #469391).
  * minor cleanups in debian/rules for new upstream release.
  * don't ignore make clean in clean target of debian rules. Thanks to
    lintian.

src/keepassx.h
/***************************************************************************
 *   Copyright (C) 2005-2008 by Tarek Saidi                                *
 *   tarek.saidi@arcor.de                                                  *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; version 2 of the License.               *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef KEEPASSX_H_
#define KEEPASSX_H_

// global defines

#define APP_DISPLAY_NAME    "KeePassX"
#define APP_CODE_NAME       "keepassx"
#define APP_SHORT_FUNC      "Password Manager"
#define APP_LONG_FUNC       "Cross Platform Password Manager"
#define APP_VERSION         "0.4.3-4"

#define BUILTIN_ICONS 69

#define KEEPASS_LITTLE_ENDIAN
//#define KEEPASS_BIG_ENDIAN

// often used, stable header files for pch

#ifdef __cplusplus

#include <cstdlib>
#include <cmath>

#include <QApplication>
#include <QBitArray>
#include <QByteArray>
#include <QCheckBox>
#include <QColor>
#include <QComboBox>
#include <QDateTime>
#include <QDialog>
#include <QDir>
#include <QDomElement>
#include <QFile>
#include <QHash>
#include <QIcon>
#include <QImage>
#include <QLabel>
#include <QLineEdit>
#include <QList>
#include <QMenu>
#include <QMessageBox>
#include <QPainter>
#include <QPaintEvent>
#include <QPixmap>
#include <QPushButton>
#include <QSettings>
#include <QString>
#include <QStringList>
#include <QTextCodec>
#include <QTimer>
#include <QTreeWidget>
#include <QUrl>

#include "crypto/aescpp.h"
#include "crypto/yarrow.h"
#include "lib/bookmarks.h"
#include "lib/FileDialogs.h"
#include "lib/SecString.h"
#include "lib/tools.h"
#include "lib/UrlLabel.h"
#include "Database.h"
#include "KpxConfig.h"

#include "plugins/interfaces/IIconTheme.h"

extern QString PluginLoadError;
extern IIconTheme* IconLoader;
extern KpxConfig *config;
extern QString AppDir;
extern QString HomeDir;
extern QString DataDir;
extern bool TrActive;
extern QString DetailViewTemplate;
extern QPixmap *EntryIcons;
extern bool EventOccurred;
extern bool EventOccurredBlock;

#endif //__cplusplus
#endif //KEEPASS_X_


src/mainwindow.h
/***************************************************************************
 *   Copyright (C) 2005-2006 by Tarek Saidi                                *
 *   tarek.saidi@arcor.de                                                  *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; version 2 of the License.               *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/


#ifndef MAINWINDOW_H
#define MAINWINDOW_H

#include <QSessionManager>
#include <QSystemTrayIcon>
#include "lib/EntryView.h"
#include "lib/GroupView.h"
#include "export/Export.h"

#include "ui_MainWindow.h"
#include "ui_WorkspaceLockedWidget.h"

class KeepassMainWindow : public QMainWindow, private Ui_MainWindow{
	Q_OBJECT
	public:
		KeepassMainWindow (const QString& ArgFile,bool ArgMin,bool ArgLock,QWidget *parent=0, Qt::WFlags flags=0);
		IDatabase* db;
		inline bool isLocked() { return IsLocked; };
		inline bool isOpened() { return FileOpen; };
	
	public slots:
		void OnUnLockWorkspace();
		void openFile(const QString& filename);

	signals:
		void entryChanged();

	private slots:
		void OnFileNewKdb();
		//void OnFileNewKxdb(); // TODO Kxdb
		void OnFileOpen();
		void OnFileClose();
		bool OnFileSave();
		bool OnFileSaveAs();
		void OnFileSettings();
		void OnFileChangeKey();
		void OnFileExit();
		void OnBookmarkTriggered(QAction* action);
		void OnSearch();
		void OnGroupSearch();
		void OnViewShowToolbar(bool);
		void OnViewShowEntryDetails(bool);
		void OnViewToolbarIconSize16(bool);
		void OnViewToolbarIconSize22(bool);
		void OnViewToolbarIconSize28(bool);
		void OnGroupSelectionChanged(IGroupHandle*);
		void OnQuickSearch();
		void OnColumnVisibilityChanged();
		void OnUsernPasswVisibilityChanged();
		void OnFileModified();
		void OnExtrasSettings();
		void OnExtrasPasswordGen();
		void OnExtrasShowExpiredEntries();
		//void OnExtrasTrashCan(); //TODO TrashCan
		void OnHelpAbout();
		void OnHelpHandbook();
		//void OnItemExpanded(QTreeWidgetItem*);
		//void OnItemCollaped(QTreeWidgetItem*);
		void OnShowSearchResults();
		void OnEntryChanged(SelectionState);
		void OnSysTrayActivated(QSystemTrayIcon::ActivationReason);
		void restoreWindow();
		void OnImport(QAction*);
		void OnExport(QAction*);
		void OnDetailViewUrlClicked(const QUrl& url);
		void OnLockClose();
		void OnInactivityTimer();
		void OnShutdown(QSessionManager& manager);
		void loadColumnVisibility();

	private:
		void closeEvent(QCloseEvent* event);
		void hideEvent(QHideEvent* event);
		void showEvent(QShowEvent* event);
		void setLock();
		void resetLock();
		enum StatusBarMsg {
			StatusBarReady, StatusBarLoading, StatusBarLoadingFailed,
			StatusBarReadOnlyLock, StatusBarReadOnly
		};
		void setStatusBarMsg(StatusBarMsg statusBarMsg);
		SelectionState GroupSelection, EntrySelection;
		bool FileOpen;
		bool ModFlag;
		bool IsLocked;
		QList<quint32> SearchResults;
		inline void setupToolbar();
		inline void setupIcons();
		inline void setupConnections();
		inline void setupMenus();
		void setStateFileOpen(bool);
		void setStateFileModified(bool);
		void setStateGroupSelected(SelectionState s);
		void setStateEntrySelected(SelectionState s);
		bool openDatabase(QString filename,bool IsAuto=false);
		void fakeOpenDatabase(const QString& filename);
		void setupDatabaseConnections(IDatabase* DB);
		bool closeDatabase(bool lock=false);
		void search(IGroupHandle* Group);
		//void removeFromSearchResults(int sID);
		void updateDetailView();
		void exportDatabase(IExport* exporter,QStringList filters);
		void saveLastFilename(const QString& filename);
		void createBookmarkActions();
		void updateWindowTitle();
		void updateTrayTooltip();
		void updateCurrentFile(const QString& filePath);
		QLineEdit* QuickSearchEdit;
		QLabel* StatusBarGeneral;
		//QLabel* StatusBarSelection;
		QToolBar* toolBar;
		QSystemTrayIcon* SysTray;
		QAction* ViewShowToolbarAction;
		QMenu* SysTrayMenu;
		//QAssistantClient* HelpBrowser; //TODO HelpBrowser
		QWidget* NormalCentralWidget;
		QWidget* LockedCentralWidget;
		Ui_WorkspaceLockedWidget WorkspaceLockedWidget;
		bool ShutingDown;
		bool InUnLock;
		QList<int> lockGroup;
		QDialog* unlockDlg;
		QString currentFilePath;
		QString currentFileName;
		int inactivityCounter;
		QTimer* inactivityTimer;
		StatusBarMsg statusbarState;
		bool dbReadOnly;
};

#endif


src/dialogs/PasswordDlg.h
/***************************************************************************
 *   Copyright (C) 2005-2007 by Tarek Saidi                                *
 *   tarek.saidi@arcor.de                                                  *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; version 2 of the License.               *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef PASSWORDDIALOG_H
#define PASSWORDDIALOG_H

#include <QPaintEvent>
#include "ui_PasswordDlg.h"

#define PASSWORDDIALOG_FLAG_NONE		0
#define PASSWORDDIALOG_FLAG_AUTO		1 << 0	// Dialog was automatically opened on start-up, currently unused
#define PASSWORDDIALOG_FLAG_OPENREADONLY	1 << 1	// Ask user to open database in read-only mode

class PasswordDialog : public QDialog, private Ui_PasswordDlg {
	Q_OBJECT
	public:
		enum DlgMode {
			Mode_Ask,     // Normal password entry when opening a database
			Mode_Set,     // Setting password for the first time after creating a new database
			Mode_Change   // Changing the password of a database
		};
		
		enum DlgExit {
			Exit_Ok=QDialog::Accepted,
			Exit_Cancel=QDialog::Rejected,
			Exit_Quit=3
		};
		
		typedef bool (KeyFileGenProc)(const QString& filename,QString* error);
			
		PasswordDialog(QWidget* parent,DlgMode mode,int flags,const QString& filename=QString());
		
		// result functions
		QString selectedBookmark();
		QString keyFile();
		QString password();
		bool openReadOnly();
	
	private slots:
		void OnOK();
		void OnCancel();
		void OnButtonBrowse();
		void OnButtonQuit();
		void OnGenKeyFile();
		void OnButtonBack();
		void ChangeEchoModeDatabaseKey();
		void OnBookmarkTriggered(QAction* action);
		void OnCheckBoxesChanged();
		void done(int r);

	private:
		DlgMode Mode;
		QPixmap BannerPixmap;
		QString BookmarkFilename;
		QString Filename;
		QString Password;
		QString KeyFile;
		void setStatePasswordOnly();
		void setStateKeyFileOnly();
		void setStateBoth();
		virtual void paintEvent(QPaintEvent*);
};

#endif



src/KpxConfig.h
/***************************************************************************
 *   Copyright (C) 2005-2007 by Tarek Saidi                                *
 *   tarek.saidi@arcor.de                                                  *
 *                                                                         *
 *   Copyright (C) 2007 by Constantin "Dinosaur" Makshin                   *
 *   dinosaur-rus@users.sourceforge.net                                    *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; version 2 of the License.               *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef _KPXCONFIG_H_
#define _KPXCONFIG_H_

#include "lib/AutoType.h"

#if defined(Q_WS_MAC)
#	define DEFAULT_MOUNT_DIR "/Volumes/"
#elif defined(Q_WS_X11)
#	define DEFAULT_MOUNT_DIR "/media/"
#elif defined(Q_WS_WIN)
#	define DEFAULT_MOUNT_DIR "/"
#else
#	define DEFAULT_MOUNT_DIR QString()
#endif

class KpxConfig{
public:
	friend class KpxBookmarks;
	
	enum IntegrPluginType{NoIntegr,KDE,Gnome};
	enum GrpTreeState{DoNothing,RestoreLast,ExpandAll};

	KpxConfig(const QString& filePath);

	bool alternatingRowColors(){return settings.value("Options/AlternatingRowColors",true).toBool();}
	QColor bannerColor1(){return stringToColor(settings.value("Options/BannerColor1","0,85,127").toString());}
	QColor bannerColor2(){return stringToColor(settings.value("Options/BannerColor2","0,117,175").toString());}
	QColor bannerTextColor(){return stringToColor(settings.value("Options/BannerTextColor","222,222,222").toString());}
	int clipboardTimeOut(){return settings.value("Options/ClipboardTimeOut",20).toInt();}
	QBitArray columns(){return stringToBitArray(settings.value("UI/Columns","11111000000").toString(),11);}
	QList<int> columnOrder(){return stringToIntArray(settings.value("UI/ColumnOrder","1,2,3,4,5,6,7,8,9,10,11").toString(),11);}
	QList<int> columnSizes(){return stringToIntArray(settings.value("UI/ColumnSizes","15,10,10,10,10,10,10,10,10,10,10").toString(),11);}
	int columnSort(){return settings.value("UI/ColumnSort",0).toInt();}
	Qt::SortOrder columnSortOrder(){return static_cast<Qt::SortOrder>(settings.value("UI/ColumnSortOrder",Qt::AscendingOrder).toInt());}
	QBitArray searchColumns(){return stringToBitArray(settings.value("UI/SearchColumns","11110000001").toString(),11);}
	QList<int> searchColumnOrder(){return stringToIntArray(settings.value("UI/SearchColumnOrder","2,3,4,5,6,7,8,9,10,11,1").toString(),11);}
	QList<int> searchColumnSizes(){return stringToIntArray(settings.value("UI/SearchColumnSizes","15,10,10,10,10,10,10,10,10,10,10").toString(),11);}
	int searchColumnSort(){return settings.value("UI/SearchColumnSort",0).toInt();}
	Qt::SortOrder searchColumnSortOrder(){return static_cast<Qt::SortOrder>(settings.value("UI/SearchColumnSortOrder",Qt::AscendingOrder).toInt());}
	QStringList fileDlgHistory(unsigned index){return settings.value(QString("FileDlgHistory/ENTRY%1").arg(index)).toStringList();}
	GrpTreeState groupTreeState(){return stringToGrpTreeState(settings.value("Options/GroupTreeState").toString());}
	bool hidePasswords(){return settings.value("UI/HidePasswords",true).toBool();}
	bool hideUsernames(){return settings.value("UI/HideUsernames",true).toBool();}
	QByteArray hSplitterPos(){return settings.value("UI/HSplitterPos").toByteArray();}
	bool alwaysOnTop(){return settings.value("UI/AlwaysOnTop",false).toBool();}
	IntegrPluginType integrPlugin(){return stringToIntegrPluginType(settings.value("Options/IntegrPlugin").toString());}
	QString lastFile(){return settings.value("Options/LastFile").toString();}
	QString lastKeyLocation(){return settings.value("Options/LastKeyLocation").toString();}
	tKeyType lastKeyType(){return stringToKeyType(settings.value("Options/LastKeyType").toString());}
	QByteArray mainWindowGeometry();
	bool minimizeToTray(){return settings.value("Options/MinimizeToTray",false).toBool();} 	
	bool minimizeTray(){return settings.value("Options/MinimizeTray",false).toBool();}
	bool startMinimized(){return settings.value("Options/StartMinimized",false).toBool();}
	bool startLocked(){return settings.value("Options/StartLocked",false).toBool();}
	QString mountDir(){return settings.value("Options/MountDir",DEFAULT_MOUNT_DIR).toString();}
	bool openLastFile(){return settings.value("Options/OpenLastFile",true).toBool();}
	bool autoSave(){return settings.value("Options/AutoSave",false).toBool();}
	bool autoSaveChange(){return settings.value("Options/AutoSaveChange",false).toBool();}
	bool backup(){return settings.value("Options/Backup",true).toBool();}
	bool backupDelete(){return settings.value("Options/BackupDelete",false).toBool();}
	int backupDeleteAfter(){return settings.value("Options/BackupDeleteAfter",14).toInt();}
	int pwGenCategory(){return settings.value("Options/PwGenCategory",0).toInt();}
	QString pwGenCharList(){return settings.value("Options/PwGenCharList").toString();}
	bool pwGenExcludeLookAlike(){return settings.value("Options/PwGenExcludeLookAlike").toBool();}
	bool pwGenEveryGroup(){return settings.value("Options/PwGenEveryGroup").toBool();}
	int pwGenLength(){return settings.value("Options/PwGenLength",25).toInt();}
	QBitArray pwGenOptions(){return stringToBitArray(settings.value("Options/PwGenOptions","11111000011110").toString(),14);}
	bool rememberLastKey(){return settings.value("Options/RememberLastKey",true).toBool();}
	bool saveFileDlgHistory(){return settings.value("Options/SaveFileDlgHistory",true).toBool();}
	bool saveRelativePaths(){return settings.value("Options/SaveRelativePaths",true).toBool();}
	QBitArray searchOptions(){return stringToBitArray(settings.value("Options/SearchOptions","001101111").toString(),9);}
	bool showEntryDetails(){return settings.value("UI/ShowEntryDetails",true).toBool();}
	bool showPasswords(){return settings.value("Options/ShowPasswords",false).toBool();}
	bool showPasswordsPasswordDlg(){return settings.value("Options/ShowPasswordsPasswordDlg",false).toBool();}
	bool lockOnMinimize(){return settings.value("Options/LockOnMinimize",false).toBool();}
	bool lockOnInactivity(){return settings.value("Options/LockOnInactivity",false).toBool();}
	int lockAfterSec(){return settings.value("Options/LockAfterSec",30).toInt();}
	bool showStatusbar(){return settings.value("UI/ShowStatusbar",true).toBool();}
	bool showSysTrayIcon(){return settings.value("Options/ShowSysTrayIcon",false).toBool();}
	bool showToolbar(){return settings.value("UI/ShowToolbar",true).toBool();}
	int toolbarIconSize(){return settings.value("UI/ToolbarIconSize",16).toInt();}
	QString urlCmd(){return settings.value("Options/UrlCmd").toString();}
	bool urlCmdDef(){return settings.value("Options/UrlCmdDef",true).toBool();}
	QByteArray vSplitterPos(){return settings.value("UI/VSplitterPos").toByteArray();}
	bool askBeforeDelete(){return settings.value("Options/AskBeforeDelete",true).toBool();}
#ifdef AUTOTYPE
	int autoTypePreGap(){return settings.value("Options/AutoTypePreGap",500).toInt();}
	int autoTypeKeyStrokeDelay(){return settings.value("Options/AutoTypeKeyStrokeDelay",5).toInt();}
#endif
#ifdef GLOBAL_AUTOTYPE
	Shortcut globalShortcut();
	bool entryTitlesMatch(){return settings.value("Options/EntryTitlesMatch",true).toBool();}
#endif
	//bool featureBookmarks(){return settings.value("Features/Bookmarks",true).toBool();}
	bool featureBookmarks(){return true;}
	QString language(){return settings.value("Options/Language","auto").toString();}
	bool openReadOnly(){return settings.value("Options/OpenReadOnly",false).toBool();}
	
	void setAlternatingRowColors(bool value){settings.setValue("Options/AlternatingRowColors",value);}
	void setBannerColor1(const QColor& value){settings.setValue("Options/BannerColor1",colorToString(value));}
	void setBannerColor2(const QColor& value){settings.setValue("Options/BannerColor2",colorToString(value));}
	void setBannerTextColor(const QColor& value){settings.setValue("Options/BannerTextColor",colorToString(value));}
	void setClipboardTimeOut(int value){settings.setValue("Options/ClipboardTimeOut",value);}
	void setColumns(const QBitArray& value){settings.setValue("UI/Columns",bitArrayToString(value));}
	void setColumnOrder(const QList<int>& value){settings.setValue("UI/ColumnOrder",intArrayToString(value));}
	void setColumnSizes(const QList<int>& value){settings.setValue("UI/ColumnSizes",intArrayToString(value));}
	void setColumnSort(int value){settings.setValue("UI/ColumnSort",value);}
	void setColumnSortOrder(int value){settings.setValue("UI/ColumnSortOrder",value);}
	void setSearchColumns(const QBitArray& value){settings.setValue("UI/SearchColumns",bitArrayToString(value));}
	void setSearchColumnOrder(const QList<int>& value){settings.setValue("UI/SearchColumnOrder",intArrayToString(value));}
	void setSearchColumnSizes(const QList<int>& value){settings.setValue("UI/SearchColumnSizes",intArrayToString(value));}
	void setSearchColumnSort(int value){settings.setValue("UI/SearchColumnSort",value);}
	void setSearchColumnSortOrder(int value){settings.setValue("UI/SearchColumnSortOrder",value);}
	void setFileDlgHistory(unsigned index,const QStringList& value){settings.setValue(QString("FileDlgHistory/ENTRY%1").arg(index), value);}
	void setGroupTreeState(GrpTreeState value){settings.setValue("Options/GroupTreeState",grpTreeStateToString(value));}
	void setHidePasswords(bool value){settings.setValue("UI/HidePasswords",value);}
	void setHideUsernames(bool value){settings.setValue("UI/HideUsernames",value);}
	void setHSplitterPos(const QByteArray& value){settings.setValue("UI/HSplitterPos",value);}
	void setAlwaysOnTop(bool value){settings.setValue("UI/AlwaysOnTop",value);}
	void setIntegrPlugin(IntegrPluginType value){settings.setValue("Options/IntegrPlugin",integrPluginTypeToString(value));}
	void setLastFile(const QString& value){settings.setValue("Options/LastFile",value);}
	void setLastKeyLocation(const QString& value){settings.setValue("Options/LastKeyLocation",value);}
	void setLastKeyType(tKeyType value){settings.setValue("Options/LastKeyType",keyTypeToString(value));}
	void setMainWindowGeometry(const QByteArray& value){settings.setValue("UI/MainWindowGeometry",value);}
	void setMinimizeToTray(bool value){settings.setValue("Options/MinimizeToTray",value);}
	void setMinimizeTray(bool value){settings.setValue("Options/MinimizeTray",value);}
	void setStartMinimized(bool value){settings.setValue("Options/StartMinimized",value);}
	void setStartLocked(bool value){settings.setValue("Options/StartLocked",value);}
	void setMountDir(const QString& value){settings.setValue("Options/MountDir",value);}
	void setOpenLastFile(bool value){settings.setValue("Options/OpenLastFile",value);}
	void setAutoSave(bool value){settings.setValue("Options/AutoSave",value);}
	void setAutoSaveChange(bool value){settings.setValue("Options/AutoSaveChange",value);}
	void setBackup(bool value){settings.setValue("Options/Backup",value);}
	void setBackupDelete(bool value){settings.setValue("Options/BackupDelete",value);}
	void setBackupDeleteAfter(int value){settings.setValue("Options/BackupDeleteAfter",value);}
	void setPwGenCategory(int value){settings.setValue("Options/PwGenCategory",value);}
	void setPwGenCharList(const QString& value){settings.setValue("Options/PwGenCharList",value);}
	void setPwGenExcludeLookAlike(bool value){settings.setValue("Options/PwGenExcludeLookAlike",value);}
	void setPwGenEveryGroup(bool value){settings.setValue("Options/PwGenEveryGroup",value);}
	void setPwGenLength(int value){settings.setValue("Options/PwGenLength",value);}
	void setPwGenOptions(const QBitArray& value){settings.setValue("Options/PwGenOptions",bitArrayToString(value));}
	void setRememberLastKey(bool value){settings.setValue("Options/RememberLastKey",value);}
	void setSaveFileDlgHistory(bool value){settings.setValue("Options/SaveFileDlgHistory",value);}
	void setSaveRelativePaths(bool value){settings.setValue("Options/SaveRelativePaths",value);}
	void setSearchOptions(const QBitArray& value){settings.setValue("Options/SearchOptions",bitArrayToString(value));}
	void setShowEntryDetails(bool value){settings.setValue("UI/ShowEntryDetails",value);}
	void setShowPasswords(bool value){settings.setValue("Options/ShowPasswords",value);}
	void setShowPasswordsPasswordDlg(bool value){settings.setValue("Options/ShowPasswordsPasswordDlg",value);}
	void setLockOnMinimize(bool value){settings.setValue("Options/LockOnMinimize",value);}
	void setLockOnInactivity(bool value){settings.setValue("Options/LockOnInactivity",value);}
	void setLockAfterSec(int value){settings.setValue("Options/LockAfterSec",value);}
	void setShowStatusbar(bool value){settings.setValue("UI/ShowStatusbar",value);}
	void setShowSysTrayIcon(bool value){settings.setValue("Options/ShowSysTrayIcon",value);}
	void setShowToolbar(bool value){settings.setValue("UI/ShowToolbar",value);}
	void setToolbarIconSize(int value){settings.setValue("UI/ToolbarIconSize",value);}
	void setUrlCmd(const QString& value){settings.setValue("Options/UrlCmd",value);}
	void setUrlCmdDef(bool value){settings.setValue("Options/UrlCmdDef",value);}
	void setVSplitterPos(const QByteArray& value){settings.setValue("UI/VSplitterPos",value);}
	void setAskBeforeDelete(bool value){settings.setValue("Options/AskBeforeDelete",value);}
#ifdef AUTOTYPE
	void setAutoTypePreGap(int value){settings.setValue("Options/AutoTypePreGap",value);}
	void setAutoTypeKeyStrokeDelay(int value){settings.setValue("Options/AutoTypeKeyStrokeDelay",value);}
#endif
#ifdef GLOBAL_AUTOTYPE
	void setGlobalShortcut(const Shortcut& s);
	void setEntryTitlesMatch(bool value){settings.setValue("Options/EntryTitlesMatch",value);}
#endif
	//void setFeatureBookmarks(bool value){settings.setValue("Features/Bookmarks",value);}
	void setLanguage(const QString& value){settings.setValue("Options/Language",value);}
	void setOpenReadOnly(bool value){settings.setValue("Options/OpenReadOnly",value);}

	unsigned fileDlgHistorySize();
	void clearFileDlgHistory(){settings.remove("FileDlgHistory");};
	
	QRect dialogGeometry(const QWidget* widget);
	void setDialogGeometry(const QWidget* widget);
	
	QString detailViewTemplate();
	QString defaultDetailViewTemplate();
	void setDetailViewTemplate(const QString& value);

private:
	QSettings settings;
	QString configFile;

	QColor stringToColor(const QString& str);
	QBitArray stringToBitArray(const QString& str, unsigned count);
	GrpTreeState stringToGrpTreeState(const QString& str);
	QList<int> stringToIntArray(const QString& str, unsigned count);
	IntegrPluginType stringToIntegrPluginType(const QString& str);
	tKeyType stringToKeyType(const QString& str);

	QString colorToString(const QColor& color){return QString("%1,%2,%3").arg(color.red()).arg(color.green()).arg(color.blue());}
	QString bitArrayToString(const QBitArray& bitArray);
	QString grpTreeStateToString(GrpTreeState grpTreeState);
	QString intArrayToString(const QList<int>& intArray);
	QString integrPluginTypeToString(IntegrPluginType integrPluginType);
	QString keyTypeToString(tKeyType keyType);
};

#endif
